File: ui/src/types/partSubCategory.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";

// Part SubCategory List
export interface PartSubCategoryDetails {
    Id: number;
    Name: string;
    Code: string;
    PartCategory: string;
    ProductCategory: string;
    IsActive: boolean;
}

export const partSubCategoryDetailsDecoder: Decoder<PartSubCategoryDetails> = object({
    Id: number,
    Name: string,
    Code: string,
    PartCategory: string,
    ProductCategory: string,
    IsActive: boolean,
});
export interface PartSubCategoryList {
    PartSubCategories: PartSubCategoryDetails[],
    TotalRows: number;
    PerPage:number;
}

export const partSubCategoryListDecoder: Decoder<PartSubCategoryList> = object({
    PartSubCategories: array(partSubCategoryDetailsDecoder),
    TotalRows: number,
    PerPage:number
});

// PartSubCategory Edit
export interface PartSubCategoryEditDetails {
    Id: null | number
    Name: string;
    IsActive: boolean;
}

export interface PartSubCategoryEditResponse {
    IsPartSubCategoryUpdated: Boolean;
}

export const partSubCategoryEditDecoder: Decoder<PartSubCategoryEditResponse> = object({
    IsPartSubCategoryUpdated: boolean,
});
// PartSubCategory Edit End

export interface PartSubCategoryNames {
    Id: number;
    Name: string;
}

export const partSubCategoryNamesDecoder: Decoder<PartSubCategoryNames> = object({
    Id: number,
    Name: string,
})

export interface PartSubCategoryNameList {
    PartSubCategoryDetails: PartSubCategoryNames[]
}

export const partSubCategoryNameListDecoder: Decoder<PartSubCategoryNameList> = object({
    PartSubCategoryDetails: array(partSubCategoryNamesDecoder),
});

export interface PartSubCategoryCreate {
    ProductCategoryId: number | string;
    PartProductCategoryToPartCategoryMappingId: number | string|null;
    PartSubCategoryName: string;
}

export interface PartSubCategoryCreateResult {
    IsPartSubCategoryCreated: Boolean; 
}

export const partSubCategoryCreateResultDecoder: Decoder<PartSubCategoryCreateResult> = object({
    IsPartSubCategoryCreated: boolean,
});

export interface PartSubCategoryDeleted {
    IsDeleted: Boolean;
}

export const partsubCategoryDeletedDecoder: Decoder<PartSubCategoryDeleted> = object({
    IsDeleted: boolean,
});

File: api/BeSureApi/Models/PartSubCategory.cs
using System.ComponentModel.DataAnnotations;

namespace BeSureApi.Models
{
    public class PartSubCategory
    {
        public int Id { get; set; }
        public string Code { get; set; }
        public string Name { get; set; }
        public bool IsActive { get; set; }
        public string ProductCategory { get; set; }
        public string PartCategory { get; set; }
    }
    public class UpdatedPartSubCategory
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public bool IsActive { get; set; }
    }

    public class PartSubCategoryNames
    {
        public string Name { get; set; }
        public int Id { get; set; }
    }

    public class PartSubCategoryCreate
    {
        [Required(ErrorMessage = "validation_error_partsubcategorycreate_product_category_required")]
        public int ProductCategoryId { get; set; }
        [Required(ErrorMessage = "validation_error_partsubcategorycreate_part_category_required")]
        public int PartProductCategoryToPartCategoryMappingId { get; set; }
        [Required(ErrorMessage = "validation_error_partsubcategorycreate_name_required")]
        [StringLength(64, ErrorMessage = "validation_error_part_create_Name_max")]
        public string PartSubCategoryName { get; set; }        
    }
}